Versioning placement & compatibility directors/strategies

benjaminpetit opened the pull request:
Following #2837

## Proposed change
This PR I implemented configurable director/policies for versioning..

I implemented two basics compatibility and two basics versioning directors. it would be easy to inject custom directors in the future using dependency injection (like it should be possible for `PlacementDirector`)

For compatibility:
- Backward only compatible (default value)
- All versions compatible

For new placement:
- Always the minimum compatible version (default value)
- Always the latest compatible available version

More directors can be added later.

## Limitations
For now, the strategies are chosen at startup, the values or loaded from configuration. A future PR will allow changing these at runtime

jdom commented on src/Orleans/Configuration/GlobalConfiguration.cs:
```diff
@@ -593,6 +599,8 @@ internal GlobalConfiguration()
             ProviderConfigurations = new Dictionary<string, ProviderCategoryConfiguration>();
             GrainServiceConfigurations = new GrainServiceConfigurations();
+            DefaultVersionCompatibilityStrategy = BackwardCompatible.Singleton;
+			DefaultVersionPlacementStrategy = MinimumVersionPlacement.Singleton;
```
nit: there is weird indenting here and in the property declaration.
UPDATE: TABS!! don't want to start a religious war about tabs vs spaces, since I don't care, but I care about consistency, so let's continue to use spaces.

benjaminpetit replied:
Yeah I think my VS2017 is set to use tabs instead of spaces... will fix that, sorry

ReubenBond replied:
@jdom maybe we should check in a `.editorconfig` file or something. I'm surprised VS didn't automatically detect the norm, though.

jdom commented on src/OrleansRuntime/Catalog/Catalog.cs:
```diff
@@ -165,7 +169,8 @@ public override void GetObjectData(SerializationInfo info, StreamingContext cont
             SerializationManager serializationManager,
-            MultiClusterRegistrationStrategyManager multiClusterRegistrationStrategyManager)
+            MultiClusterRegistrationStrategyManager multiClusterRegistrationStrategyManager,
+			CachedVersionDirectorManager versionDirectorManager)
```
nit: weird indenting

jdom replied:
Indenting is still weird

jdom commented on src/Orleans/Versions/Compatibility/ICompatibilityDirector.cs:
```diff
@@ -0,0 +1,18 @@
+{
+    public interface IVersionCompatibilityDirector
+    {
+        bool IsCompatible(ushort requestedVersion, ushort actualVersion);
```
`actualVersion` is confusing, I had to look at the usage to understand what it is. Maybe `versionInUse` or something like that?

ReubenBond replied:
alternatively `loadedVersion` or `currentVersion`

jdom commented:
I'm glad the default placement strategy was changed to not always pick the latest. Neverhtless, I'd prefer that default is not the minimum, but instead by default try to use the requestedVersion. So new clients will effectively pick the new version (if the instance is not yet activated), while older clients will by default continue to talk to activate the previous version, even when all versions are compatible to each other.

benjaminpetit commented:
@jdom I think I wasn't clear in my description. And maybe there is a bug in my implementation now I think of.

Currently, the "minimum compatible version" is implemented like this:

`` availableVersions.Where(v => versionCompatibilityDirector.IsCompatible(requestedVersion, v)).Min(); ``

So it should take by default the requested version if available, if not the closest upper version. But maybe I should fallback to the smaller version like my description said.

jdom commented on src/Orleans/Configuration/GlobalConfiguration.cs:
```diff
@@ -449,6 +451,10 @@ internal void SetReminderServiceType(ReminderServiceProviderType reminderType)
 
         public string DefaultPlacementStrategy { get; set; }
 
+        public VersionCompatibilityStrategy DefaultVersionCompatibilityStrategy { get; set; }
```
why are these "Default"? Is there a way to override the default?

benjaminpetit replied:
In the future yes

sergeybykov replied:
I suggest we remove "Version" from the names to make them shorter or at least replace it with "Grain".

jdom commented:
@benjaminpetit well but `availableVersions.Where(v => versionCompatibilityDirector.IsCompatible(requestedVersion, v)).Min()` uses the versionCompatibilityDirector, so it will not use the requestVersion by default in all cases, such as when the compatibility director is the one that says that every version is compatible with every other. At that point, it will ALWAYS use the minimum version, regardless of what was the requested one.

jdom commented:
Basically the default I'd like is to activate the requested version, or if no longer available, fallback to either the closest higher compatible version, or the closest lower compatible version (either configurable or just pick one that makes sense), or if you don't like the "closest" behavior, then the fallback can be either minimum or latest at that point. Nevertheless, since we try to activate the request version, it kind of makes sense to fall back to the closest compatible version

jdom commented:
Just talked in person with Ben. I'm suggesting that the default (again, what will  be used in most scenarios, but you can select other policies or add your custom one) should be the one that tesnds to break less, since testing for back compat is extremely difficult for end users.

So imagine you have versions 1, 2 and 3 deployed. If a request comes for version 1, and there are active silos with version 1, then it should activate there, since it's the most tested scenario for the developer.
If silos with version 1 are shutting down, and a client requests version 1, then it makes sense to activate the closest higher version (since it's more likely that the next higher version will be compatible with the previous). Even though technically v3 should be compatible with v1, it's likely that the most tested scenario for the developer is that v2 is compatible to v1, and v3 is compatible to v2, but not necessarily tested the hell out of the compatibility between v1 and v3.

So in a summary, the default placement IMO should be:
- Activate the requested version if active silos are available
- Fall back to activating the closest higher compatible version
- Fall back to activating the closest lower compatible version (this one can fail if the compatibility director says that compatibility is only backwards, which is fine, and an activation exception is thrown).

sergeybykov commented:
>Activate the requested version if active silos are available

My concern is that unless this logic is applied proportionally to the fraction of silos with that version, we can easily overload those silos. In the extreme case, the last standing silo with v1 will end up processing all v1 requests even if there's plenty of v2/v3 silos in the cluster.

jdom commented:
Yes, but if you are shutting down v1, then it makes sense that you are also shutting down clients that request v1 also. The thing is that the overloading possibility applies to any placement director that only takes versions into account (regardless if is is the minimum, maximum or whatever version), it's just that IMO this is the scenario less likely to break in terms of real compatibility between versions (which I iteratate that it's extremely hard for developers to test in their apps).
If we keep this default, we could also provide a variation that takes how many silos of each are available, but I'd argue that if you are shutting down v1, you are very likely to be shutting down clients in v1 also, so this might prove to be a premature optimization.

benjaminpetit commented:
> My concern is that unless this logic is applied proportionally to the fraction of silos with that version, we can easily overload those silos

Using "Latest Version" you will have the same issue when a new silo version starts. 

> Yes, but if you are shutting down v1, then it makes sense that you are also shutting down clients that request v1 also

Not necessary. Let's say you are pushing a simple bugfix silo-side: you don't want to redeploy client in this case.

I think the best default value would be: get the requested version, or the closest compatible upper version. If not, fallback to a lowest one, like described by @jdom

When strategy change at runtime would be implemented, I expect the dev to use this rule by default, start new silo with the latest version, wait some time so the cluster stabilize, then change the strategy to "Always latest". Then shutdown previous version silo.

Using this method allow developer to run some "smoke test calls" again the new version without creating "real" new activations on the latest version. And it avoids overloading at silo start/shutdown.

benjaminpetit commented:
So I made some modifications:
- Added a new version director: place new activations on all compatible silos, regardless of the version order. For example if in your cluster you have 10 silos, 2 v2 and 8 v1, 20% of the new activation will go on v2, 80% will go on v1. This is the new default director for now
- Added some caching
- Sort he returned list of silos by the director

jdom commented on src/OrleansRuntime/Versions/CachedVersionDirectorManager.cs:
```diff
@@ -0,0 +1,64 @@
+            var compat = this.CompatibilityDirectorManager.GetDirector(ifaceId);
+            var availableVersions = this.grainTypeManager.GetAvailableVersions(ifaceId);
+
+            return new CachedVersionDirector(version, compat, availableVersions);
```
not sure I understand the need for the CachedVersionDirector class, since the CachedVersionDirectorManager is already caching the results too... Why these 2 layers of caching?
is it because why might have many matches for (ifaceId,requestVersion) but typeCode might be the differentiator, so it make sense to have director.GetSuitableVersion cached for that pair?

benjaminpetit replied:
You are correct, I did that because `typeCode` might be the differenciator.

But I think I might have over-engineered that. This level of caching is only useful if you have one interface for multiple `typeCode`. And you have to get  A LOT of Grain that share the interface for this cahing to be useful. I will just remove it.

jdom commented on src/Orleans/Versions/Placement/IPlacementVersionDirector.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace Orleans.Versions.Placement
+{
+    public interface IVersionPlacementDirector
```
>IVersionPlacementDirector [](start = 21, length = 25)

Filename doesn't match class name (it's inverted)

sergeybykov replied:
Maybe IVersionSelector instead?

benjaminpetit replied:
Seems a better name yes

jdom commented on src/Orleans/Versions/Placement/IPlacementVersionDirector.cs:
```diff
@@ -0,0 +1,20 @@
+{
+    public interface IVersionPlacementDirector
+    {
+        IReadOnlyList<ushort> GetSuitableVersion(ushort requestedVersion, IReadOnlyList<ushort> availableVersions, IVersionCompatibilityDirector versionCompatibilityDirector);
```
This API does not allow for weighted decisions... (like instead of randomly placing in all suitable silos, favor either the higher version if there is a big enough deployment rollout of that new version).
I'm fine that we don't over engineer this until we actually provide some of those directors in-the-box, but since it's an extension point, just wanted to raise awareness so we make a conscious choice not to support those for now.

jdom commented on test/Tester/HeterogeneousSilosTests/UpgradeTests/UpgradeTests.cs:
```diff
@@ -0,0 +1,92 @@
+        protected override VersionCompatibilityStrategy VersionCompatibilityStrategy => BackwardCompatible.Singleton;
+        
+        [Fact]
+        public Task AlwaysCreateActivationWithMinimumVersionTest()
```
>Test [](start = 60, length = 4)

please remove Test suffix from test methods

jdom commented on test/Tester/HeterogeneousSilosTests/UpgradeTests/UpgradeTests.cs:
```diff
@@ -0,0 +1,92 @@
+        }
+
+        [Fact]
+        public Task UpgradeProxyCallNoPendingRequestTest()
```
>UpgradeProxyCallNoPendingRequest [](start = 20, length = 32)

Can you explain a little bit more what you are asserting? especially since the assert is in external call

jdom commented on test/Tester/HeterogeneousSilosTests/UpgradeTests/UpgradeTests.cs:
```diff
@@ -0,0 +1,92 @@
+        [Fact]
+        public Task UpgradeProxyCallNoPendingRequestTest()
+        {
+            return ProxyCallNoPendingRequest(2);
```
>2 [](start = 45, length = 1)

consider using named parameters in all of these, such as ProxyCallNoPendingRequest(expectedVersion: 2)

ReubenBond commented on src/Orleans/Orleans.csproj:
```diff
@@ -445,6 +452,7 @@
   <ItemGroup>
     <None Include="project.json" />
   </ItemGroup>
+  <ItemGroup />
```
nit nit: superfluous

ReubenBond commented on src/OrleansRuntime/Catalog/Catalog.cs:
```diff
@@ -171,7 +174,8 @@ public override void GetObjectData(SerializationInfo info, StreamingContext cont
             IStreamProviderManager providerManager,
-            IServiceProvider serviceProvider)
+            IServiceProvider serviceProvider,
+			CachedVersionSelectorManager versionSelectorManager)
```
nit nit: some spacing weirdness here - looks like the second line is using tabs instead of spaces.

ReubenBond commented on src/OrleansRuntime/Silo/Silo.cs:
```diff
@@ -280,6 +285,16 @@ internal Silo(SiloInitializationParameters initializationParams)
             services.AddSingleton<ClientObserversPlacementDirector>();
 
+            // Versions
+            services.AddSingleton<VersionSelectorManager>();
```
How can this be constructed using DI? It has two public constructors

ReubenBond replied:
It does work, so I suppose it's fine. Did the behaviour of DI construction change in `Microsoft.Extensions.DependencyInjection` v1.1.0?

benjaminpetit replied:
I suppose it takes the first and it works... I will merge the ctor, not needed to have two here anyway

ReubenBond commented on src/OrleansRuntime/Versions/CachedVersionSelectorManager.cs:
```diff
@@ -0,0 +1,54 @@
+        public IReadOnlyList<SiloAddress> GetSuitableSilos(int typeCode, int ifaceId, ushort requestedVersion)
+        {
+            var key = Tuple.Create(typeCode, ifaceId, requestedVersion);
+            return suitableSilosCache.GetOrAdd(key, getSilosFunc);
```
It doesn't have a GetOrAdd method yet, but you could use CachedReadConcurrentDictionary instead here and have cheaper access times while retaining thread safety (and no need to reassign the dictionary when resetting the cache.

`ResetCache` below could use `.Clear()`

benjaminpetit replied:
Didn't saw this class earlier, you added recently. I will look into that

ReubenBond commented on src/OrleansRuntime/GrainTypeManager/GrainTypeManager.cs:
```diff
@@ -151,24 +149,27 @@ internal IReadOnlyList<SiloAddress> GetSupportedSilos(int typeCode)
-                    return supportedSilosByInterface[ifaceId][version].Intersect(silosWithTypeCode).ToList();
-                });
+            var result = new List<SiloAddress>();
+            foreach (var version in versions)
```
I'm sure I sound like a broken record, always complaining about allocations, but if `versions` used the concrete type instead of an interface, then this might not allocate an enumerator.

benjaminpetit replied:
I agree but I would argue that we call this method only once per type/iface/version since it is cached at an upper level. I prefer to use a strict param since no modification should be done to that list.

ReubenBond commented on src/OrleansRuntime/GrainTypeManager/GrainTypeManager.cs:
```diff
@@ -151,24 +149,27 @@ internal IReadOnlyList<SiloAddress> GetSupportedSilos(int typeCode)
+            foreach (var version in versions)
+            {
+                var silosWithTypeCode = supportedSilosByTypeCode[typeCode];
+                var silosWithCorrectVersion = supportedSilosByInterface[ifaceId][version].Intersect(silosWithTypeCode).ToList();
```
Is there a reason to call ToList if passing it to AddRange anyway?

ReubenBond commented on src/Orleans/Utils/CachedReadConcurrentDictionary.cs:
```diff
@@ -133,6 +134,25 @@ public void Add(TKey key, TValue value)
+                return value;
+
+            value = valueFactory(key);
+            Add(key, value);
```
This should be a TryAdd, and if the TryAdd fails then it should call TryGet and return the result

ReubenBond commented:
LGTM with one comment - the current GetOrAdd implementation can throw under concurrent access
